File: advent_of_code_2023_q3.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>
#include <algorithm>


void part1() {
    std::ifstream file("input.txt");
    std::string line;
    int sum = 0;
    std::vector<std::vector<char>> grid;
    while (std::getline(file, line)) {
        std::vector<char> row;
        for (int i = 0; i < line.length(); i++) {
            row.push_back(line[i]);
        }
        grid.push_back(row);
    }
    for (int row_ind = 0; row_ind < grid.size(); row_ind++) {
        int col_ind = 0;
        bool num_flag = false;
        int num_start;
        while (col_ind < grid[row_ind].size() + 1) {
            if (col_ind < grid[row_ind].size() && isdigit(grid[row_ind][col_ind]) && !num_flag){
                num_start = col_ind;
                num_flag = true;
            }
            else if (num_flag && ((col_ind < grid[row_ind].size() && !isdigit(grid[row_ind][col_ind])) || (col_ind == grid[row_ind].size()))) {
                int num_end = col_ind;
                num_flag = false;
                for (int i = num_start; i < num_end; i++) {
                    for (int adj_row = std::max(0, row_ind - 1); adj_row <= std::min(row_ind + 1, (int)grid.size() - 1); adj_row++) {
                        for (int adj_col = std::max(0, i - 1); adj_col <= std::min(i + 1, (int)grid[adj_row].size() - 1); adj_col++) {
                            if (!isdigit(grid[adj_row][adj_col]) && (grid[adj_row][adj_col] != '.')) {
                                std::string num_str = "";
                                for (int i = num_start; i < num_end; i++) {
                                    num_str += grid[row_ind][i];
                                }
                                sum += std::stoi(num_str);
                                i = num_end + 1;
                                adj_row = row_ind + 1;
                                adj_col = i + 1;
                            }
                        }
                    }
                }
            }
            col_ind++;
        }
    }
    std::cout << sum << std::endl;
}


void part2() {
    std::ifstream file("input.txt");
    std::string line;
    int sum = 0;
    std::vector<std::vector<char>> grid;
    while (std::getline(file, line)) {
        std::vector<char> row;
        for (int i = 0; i < line.length(); i++) {
            row.push_back(line[i]);
        }
        grid.push_back(row);
    }
    int row_ind = 0;
    while (row_ind < grid.size()) {
        int col_ind = 0;
        std::vector<char> row = grid[row_ind];
        while (col_ind < row.size()) {
            if (row[col_ind] == '*') {
                std::vector<int> adj_nums;
                for (int adj_row = std::max(0, row_ind - 1); adj_row <= std::min(row_ind + 1, (int)grid.size() - 1); adj_row++) {
                    int i = 0;
                    std::vector<char> adj_row_vec = grid[adj_row];
                    bool num_flag = false;
                    int num_start;
                    while (i < adj_row_vec.size() + 1) {
                        if (i < adj_row_vec.size() && isdigit(adj_row_vec[i]) && !num_flag){
                            num_start = i;
                            num_flag = true;
                        }
                        else if (num_flag && ((i < adj_row_vec.size() && !isdigit(adj_row_vec[i])) || (i == adj_row_vec.size()))) {
                            int num_end = i;
                            num_flag = false;
                            if ((num_start <= col_ind + 1) && (num_end >= col_ind)) {
                                int num_length = num_end - num_start;
                                std::string num_str = "";
                                for (int j = num_start; j < num_end; j++) {
                                    num_str += adj_row_vec[j];
                                }
                                int num = std::stoi(num_str);
                                adj_nums.push_back(num);
                                if (adj_nums.size() > 2) {
                                    i = adj_row_vec.size() + 2;
                                    adj_row = row_ind + 2;
                                }
                            }
                        }
                        i++;
                    }
                }
                if (adj_nums.size() == 2) {
                    sum += adj_nums[0] * adj_nums[1];
                }
            }
            col_ind++;
        }
        row_ind++;
    }
    std::cout << sum << std::endl;
}


int main() {
    part1();
    part2();
    return 0;
}

File: advent_of_code_2023_q3.py
f = open("input.txt", 'r')
out = f.read().split('\n')

def part1():
    lines = [list(line) for line in out]
    total = 0
    for line_ind, line in enumerate(lines):
        i = 0
        num = False
        beg_ind = -1
        curr_num = ''
        while i < len(line) + 1:
            if i == len(line) or not line[i].isdigit():
                if num:
                    num = False
                    int_num = int(curr_num)
                    j = beg_ind
                    while j < i:
                        adj_row = -1
                        while adj_row < 2:
                            adj_col = -1
                            while adj_col < 2:
                                if 0 <= line_ind + adj_row < len(lines) and 0 <= j + adj_col < len(line[0]):
                                    if lines[line_ind + adj_row][j + adj_col] != '.' and not lines[line_ind + adj_row][j + adj_col].isdigit():
                                        print(int_num)
                                        total += int_num
                                        adj_col = 2
                                        adj_row = 2
                                        j = i
                                adj_col += 1
                            adj_row += 1
                        j += 1
                    curr_num = ''
            else:
                if not num:
                    beg_ind = i
                    num = True
                curr_num += line[i]
            i += 1
    print(total)
    
    


def part2():
    lines = [list(line) for line in out]
    for line in lines:
        i = 0
        num = False
        beg_ind = -1
        curr_num = ''
        while i < len(line):
            if line[i].isdigit():
                if not num:
                    beg_ind = i
                    num = True
                curr_num += line[i]
            else:
                if num:
                    num = False
                    for j in range(beg_ind, i):
                        line[j] = int(curr_num)
                    curr_num = ''
            i += 1
        if num:
            for j in range(beg_ind, i):
                line[j] = int(curr_num)
            curr_num = ''


    m = len(lines)
    n = len(lines[0])
    total = 0
    for i in range(m):
        for j in range(n):
            if lines[i][j] == '*':
                adj_squares = [(x, y) for x in range(-1, 2) for y in range(-1, 2) if 0 <= i + x < m and 0 <= j + y < n and (x, y) != (0, 0)]
                nums = []
                prev_sq = False
                for x, y in adj_squares:
                    if type(lines[i + x][j + y]) == int:
                        if not prev_sq:
                            nums.append(lines[i + x][j + y])
                            if (x, y) in [(-1, -1), (-1, 0), (1, -1), (1, 0)]:
                                prev_sq = True
                    else:
                        prev_sq = False
                if len(nums) == 2:
                    total += nums[0] * nums[1]
                
    print(total)

part1()
part2()